File: acorde-app/core/tests/Sources.test.ts
import { LaCuerdaSource } from '../sources/LaCuerdaSource';
import { UltimateGuitarSource } from '../sources/UltimateGuitarSource';
import { CifrasSource } from '../sources/CifrasSource';

global.fetch = jest.fn();

describe('Additional Sources', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('LaCuerdaSource', () => {
    const source = new LaCuerdaSource();
    
    it('should parse search results correctly', async () => {
      const mockHtml = '<html><body><a href="/tabs/l/legiao_urbana/tempo_perdido.shtml">Tempo Perdido</a></body></html>';
      (global.fetch as jest.Mock).mockResolvedValue({
        ok: true,
        text: () => Promise.resolve(mockHtml),
      });

      const results = await source.search('tempo');
      expect(results.length).toBeGreaterThan(0);
      expect(results[0].url).toContain('lacuerda.net/tabs/');
    });

    it('should extract song content correctly', async () => {
      const mockHtml = '<html><body><h1>Song Title</h1><h2>Artist Name</h2><pre>Chords and Lyrics</pre></body></html>';
      (global.fetch as jest.Mock).mockResolvedValue({
        ok: true,
        text: () => Promise.resolve(mockHtml),
      });

      const song = await source.getSong('some-url');
      expect(song.title).toBe('Song Title');
      expect(song.artist).toBe('Artist Name');
      expect(song.lyrics).toBe('Chords and Lyrics');
    });
  });

  describe('UltimateGuitarSource', () => {
    const source = new UltimateGuitarSource();

    it('should parse search results from JSON data-content', async () => {
      const mockData = {
        store: {
          page: {
            data: {
              results: [
                { type: 'tab', song_name: 'Yellow', artist_name: 'Coldplay', tab_url: 'https://ug.com/tab1' }
              ]
            }
          }
        }
      };
      const mockHtml = `<html><body><div class="js-store" data-content='${JSON.stringify(mockData)}'></div></body></html>`;
      
      (global.fetch as jest.Mock).mockResolvedValue({
        ok: true,
        text: () => Promise.resolve(mockHtml),
      });

      const results = await source.search('yellow');
      expect(results.length).toBe(1);
      expect(results[0].title).toBe('Yellow');
      expect(results[0].artist).toBe('Coldplay');
    });
  });

  describe('CifrasSource', () => {
    const source = new CifrasSource();

    it('should parse search results correctly', async () => {
      const mockHtml = `
        <div class="search-result">
          <div class="item">
            <a href="/artist/song">
              <span class="title">Song Title</span>
              <span class="artist">Artist Name</span>
            </a>
          </div>
        </div>
      `;
      (global.fetch as jest.Mock).mockResolvedValue({
        ok: true,
        text: () => Promise.resolve(mockHtml),
      });

      const results = await source.search('test');
      expect(results.length).toBe(1);
      expect(results[0].title).toBe('Song Title');
      expect(results[0].artist).toBe('Artist Name');
    });
  });
});

File: flutter/acorde/test/core/sources_test.dart
import 'dart:convert';
import 'package:flutter_test/flutter_test.dart';
import 'package:acorde/core/sources/ultimate_guitar_source.dart';
import 'package:acorde/core/sources/cifras_source.dart';

void main() {
  group('UltimateGuitarSource', () {
    late UltimateGuitarSource source;
    String mockResponse = '';

    Future<String> mockFetchHtml(String url) async {
      return mockResponse;
    }

    setUp(() {
      source = UltimateGuitarSource(fetchHtmlFn: mockFetchHtml);
    });

    test('should parse search results from JSON data-content', () async {
      final mockData = {
        'store': {
          'page': {
            'data': {
              'results': [
                {
                  'type': 'tab',
                  'song_name': 'Yellow',
                  'artist_name': 'Coldplay',
                  'tab_url': 'https://www.ultimate-guitar.com/tab/12345',
                  'type_name': 'Chords',
                  'rating': '4.8',
                  'is_pro': false,
                },
              ],
            },
          },
        },
      };

      final mockHtml =
          '<html><body><div class="js-store" data-content=\'${jsonEncode(mockData)}\'></div></body></html>';
      mockResponse = mockHtml;

      final results = await source.search('yellow');
      expect(results.length, equals(1));
      expect(results[0].title, equals('Yellow'));
      expect(results[0].artist, equals('Coldplay'));
      expect(
        results[0].url,
        equals('https://www.ultimate-guitar.com/tab/12345'),
      );
    });
  });

  group('CifrasSource', () {
    late CifrasSource source;
    String mockResponse = '';

    Future<String> mockFetchHtml(String url) async {
      return mockResponse;
    }

    setUp(() {
      source = CifrasSource(fetchHtmlFn: mockFetchHtml);
    });

    test('should parse search results correctly', () async {
      const mockHtml = '''
        <div class="search-result">
          <div class="item">
            <a href="/artist/song">
              <span class="title">Song Title</span>
              <span class="artist">Artist Name</span>
            </a>
          </div>
        </div>
      ''';
      mockResponse = mockHtml;

      final results = await source.search('test');
      expect(results.length, equals(1));
      expect(results[0].title, equals('Song Title'));
      expect(results[0].artist, equals('Artist Name'));
    });
  });
}
